SentenceVectors.ecl

```diff
@@ -0,0 +1,651 @@
+    //  Allwords should be distributed by HASH32(text) at this point
+    //  mod should also be distributed by HASH32(text).
+    allWordsD := DISTRIBUTE(allWords, HASH32(text));
+    modD := DISTRIBUTE(mod, HASH32(text));
+    allWordsS := SORT(allWordsD, text, LOCAL);
+    // Find unique
+    words := DEDUP(allWordsS, text, LOCAL);
+    // Join the unique words with the model.  If a word is not in the model, find the closest
+    // word to it in typographic edit distance.
+    wordsM0 := JOIN(words, modD, LEFT.text = RIGHT.text,
+                TRANSFORM(mappingRec, SELF.orig := LEFT.text,
+                  SELF.new := RIGHT.text),
+               LEFT OUTER, LOCAL);
+    missing := wordsM0(LENGTH(new) = 0);
+    found := findClosestMatch(mod, missing);
+    good := wordsM0(LENGTH(new) != 0);
+    wordsM := good + found;
+    // Now map the changes back into a new version of allWords
+    mapped := JOIN(allWordsD, wordsM, LEFT.text = RIGHT.orig,
+                      TRANSFORM(WordExt, SELF.text := RIGHT.new,
+                                  SELF := LEFT), LOCAL);
+    // Re DISTRIBUTE since some of the text may have changed.
+    RETURN DISTRIBUTE(mapped, HASH32(text));
+  END;
+  /**
+    * Convert a sentence into a sentence vector based on the words in the sentence and
+    * the words in the model.
+    */
+  SHARED SentInfo sent2vector(DATASET(TextMod) mod, DATASET(Sentence) sent, UNSIGNED2 vecLen,
+                          BOOLEAN mapMissingWords = TRUE) := FUNCTION
+    // Should ultimately optimize with C++
+    corp := int.Corpus(wordNGrams := wordNGrams);
+    // Get the tokenized sentence
+    wl := corp.sent2wordList(sent);
+    wordExt getWords(WordList w, UNSIGNED c) := TRANSFORM
+      SELF.sentId := w.sentId;
+      SELF.text := w.words[c];
+    END;
+    // Create a separate record for each word
+    allWords := NORMALIZE(wl, COUNT(LEFT.words), getWords(LEFT, COUNTER));
+    allWordsD := DISTRIBUTE(allWords, HASH32(text));
+    modD := DISTRIBUTE(mod, HASH32(text));
+    // If requested, map any words not in the model to the closest approximation
+    allWordsM := IF(mapMissingWords, mapWords(modD, allWordsD), allWordsD);
+    // Get the word vectors for each word from the model
+    sentWords0 := JOIN(allWordsM, modD, LEFT.text = RIGHT.text,
+                        TRANSFORM(SentInfo, SELF.sentId := LEFT.sentId,
+                            SELF.vec := RIGHT.vec,
+                            SELF := LEFT),
+                        LOCAL);
+    // Redistribute by sentence id
+    sentWords := SORT(DISTRIBUTE(sentWords0, sentId), sentId, LOCAL);
+    SentInfo doRollup(sentInfo lr, SentInfo rr) := TRANSFORM
+      SELF.vec := lr.vec + rr.vec;
+      SELF.sentId := lr.sentId;
+      SELF.text := '';
+    END;
+    // Accumulate all of the word vectors into a single long vector
+    sentOut0 := ROLLUP(sentWords, doRollup(LEFT, RIGHT), sentId, LOCAL);
```

Comment:
Again group rollup is likely to be more efficient.

Reply:
deferred